tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp

```diff
@@ -0,0 +1,591 @@
+    }
+    else
+    {
+      throw std::runtime_error("Unsupport OFKTNode type!");
+    }
+
+    cloneHelper(cloned, child);
+  }
+}
+
+StateSolver::Ptr OFKTStateSolver::clone() const
+{
+  auto cloned = std::make_shared<OFKTStateSolver>();
+  cloned->root_ = std::make_unique<OFKTRootNode>(root_->getLinkName());
+  cloned->link_map_[root_->getLinkName()] = cloned->root_.get();
+  cloned->current_state_ = std::make_shared<EnvState>(*current_state_);
+  cloned->joint_names_ = joint_names_;
+  cloned->limits_ = limits_;
+  return cloned;
+}
+
+bool OFKTStateSolver::init(tesseract_scene_graph::SceneGraph::ConstPtr scene_graph)
+{
+  assert(scene_graph->isTree());
+
+  const std::string& root_name = scene_graph->getRoot();
+
+  root_ = std::make_unique<OFKTRootNode>(root_name);
+  link_map_[root_name] = root_.get();
+  current_state_->link_transforms[root_name] = root_->getWorldTransformation();
+
+  std::vector<std::pair<std::string, std::array<double, 2>>> limits;
+  limits.resize(scene_graph->getJoints().size());
+  ofkt_builder builder(*this, limits);
+
+  std::map<tesseract_scene_graph::SceneGraph::Vertex, size_t> index_map;
+  boost::associative_property_map<std::map<tesseract_scene_graph::SceneGraph::Vertex, size_t>> prop_index_map(
+      index_map);
+
+  int c = 0;
+  tesseract_scene_graph::Graph::vertex_iterator i, iend;
+  for (boost::tie(i, iend) = boost::vertices(*scene_graph); i != iend; ++i, ++c)
+    boost::put(prop_index_map, *i, c);
+
+  boost::depth_first_search(
+      static_cast<const tesseract_scene_graph::Graph&>(*scene_graph),
+      boost::visitor(builder).root_vertex(scene_graph->getVertex(root_name)).vertex_index_map(prop_index_map));
+
+  // Populate Joint Limits
+  limits_.resize(static_cast<long>(limits.size()), 2);
+  for (std::size_t i = 0; i < limits.size(); ++i)
+  {
+    limits_(static_cast<long>(i), 0) = limits[i].second[0];
+    limits_(static_cast<long>(i), 1) = limits[i].second[1];
+  }
+
+  revision_ = 1;
+
+  update(root_.get(), false);
+
+  return true;
+}
+
+void OFKTStateSolver::setState(const std::unordered_map<std::string, double>& joints)
+{
+  for (const auto& joint : joints)
+  {
+    nodes_[joint.first]->storeJointValue(joint.second);
+    current_state_->joints[joint.first] = joint.second;
+  }
+
+  update(root_.get(), false);
+}
+
+void OFKTStateSolver::setState(const std::vector<std::string>& joint_names, const std::vector<double>& joint_values)
+{
+  for (std::size_t i = 0; i < joint_names.size(); ++i)
+  {
+    nodes_[joint_names[i]]->storeJointValue(joint_values[i]);
+    current_state_->joints[joint_names[i]] = joint_values[i];
+  }
+
+  update(root_.get(), false);
+}
+
+void OFKTStateSolver::setState(const std::vector<std::string>& joint_names,
+                               const Eigen::Ref<const Eigen::VectorXd>& joint_values)
+{
+  for (std::size_t i = 0; i < joint_names.size(); ++i)
+  {
+    nodes_[joint_names[i]]->storeJointValue(joint_values(static_cast<long>(i)));
```

Comment:
@Levi-Armstrong I'm getting a segfault here. joint_names[i] is not in the map, so it creates one. Then ->storeJointValue is derefencing a nullptr.

Reply:
While this should probably be fixed. It looks like the cause was that clone does not populate nodes_

Reply:
It looks like cloneHelper is not being called.